server/app/services/program/predicate/PredicateGenerator.java

```diff
@@ -0,0 +1,290 @@
+      Matcher multiValueMatcher = MULTI_PREDICATE_VALUE_FORM_KEY_PATTERN.matcher(key);
+
+      int groupId;
+      long questionId;
+      Scalar scalar;
+      Operator operator;
+      PredicateValue predicateValue;
+
+      if (singleValueMatcher.find()) {
+        Pair<Integer, Long> groupIdQuestionIdPair =
+            getGroupIdAndQuestionId(roQuestionService, singleValueMatcher);
+        groupId = groupIdQuestionIdPair.getLeft();
+        questionId = groupIdQuestionIdPair.getRight();
+        Pair<Scalar, Operator> scalarOperatorPair = getScalarAndOperator(predicateForm, questionId);
+        scalar = scalarOperatorPair.getLeft();
+        operator = scalarOperatorPair.getRight();
+
+        predicateValue =
+            parsePredicateValue(scalar, operator, predicateForm.get(key), ImmutableList.of());
+      } else if (multiValueMatcher.find() && !consumedMultiValueKeys.contains(key)) {
+        Pair<Integer, Long> groupIdQuestionIdPair =
+            getGroupIdAndQuestionId(roQuestionService, multiValueMatcher);
+        groupId = groupIdQuestionIdPair.getLeft();
+        questionId = groupIdQuestionIdPair.getRight();
+        Pair<Scalar, Operator> scalarOperatorPair = getScalarAndOperator(predicateForm, questionId);
+        scalar = scalarOperatorPair.getLeft();
+        operator = scalarOperatorPair.getRight();
+
+        ImmutableList<String> multiSelectKeys =
+            predicateForm.rawData().keySet().stream()
+                .filter(
+                    filteredKey ->
+                        filteredKey.startsWith(
+                            String.format(
+                                "group-%d-question-%d-predicateValues", groupId, questionId)))
+                .collect(ImmutableList.toImmutableList());
+
+        consumedMultiValueKeys.addAll(multiSelectKeys);
+
+        ImmutableList<String> rawPredicateValues =
+            multiSelectKeys.stream()
+                .map(predicateForm.rawData()::get)
+                .collect(ImmutableList.toImmutableList());
+
+        predicateValue = parsePredicateValue(scalar, operator, "", rawPredicateValues);
+      } else {
+        continue;
+      }
+
+      leafNodes.put(
+          groupId,
+          LeafOperationExpressionNode.builder()
+              .setQuestionId(questionId)
+              .setScalar(scalar)
+              .setOperator(operator)
+              .setComparedValue(predicateValue)
+              .build());
+    }
+
+    switch (detectFormat(leafNodes)) {
+      case OR_OF_SINGLE_LAYER_ANDS:
+        {
+          return PredicateDefinition.create(
+              PredicateExpressionNode.create(
+                  OrNode.create(
+                      leafNodes.keySet().stream()
+                          .map(leafNodes::get)
+                          .map(
+                              leafNodeGroup ->
+                                  leafNodeGroup.stream()
+                                      .map(PredicateExpressionNode::create)
```

Comment:
How are we saving the admin specified question order for when we recreated the UI for editing?

Reply:
see above comment